File: my-folder/0378-kth-smallest-element-in-a-sorted-matrix/solution.js
/**
 * @param {number[][]} matrix
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function(matrix, k) {
    let minHeap = new MinPriorityQueue(v => v[0])
    let n = matrix.length
    let res = 0

    for (let i = 0; i < n; i++) {
        minHeap.enqueue([matrix[i][0], i, 0])
    }

    while (k-- > 0) {
        const [v, r, c] = minHeap.dequeue()
        res = v

        if (c+1 < n) {
            minHeap.enqueue([matrix[r][c+1], r, c+1])
        }
    }

    return res
};

File: my-folder/0378-kth-smallest-element-in-a-sorted-matrix/solution.go
func kthSmallest(matrix [][]int, k int) int {
    minHeap := priorityqueue.NewWith(utils.IntComparator)
    n := len(matrix)
    var res interface{}
    for i := 0; i < n; i++ {
        for j := 0; j < n; j++ {
            minHeap.Enqueue(matrix[i][j])
        }
    }

    for k >= 0 {
        if k == 0 {
           break
        }
        res, _ = minHeap.Dequeue()
        k--
    }
    return res.(int)
}
